curator-framework/src/test/java/org/apache/curator/framework/imps/TestWatcherIdentity.java

```diff
@@ -32,167 +38,152 @@
+			client.create().forPath("/test");
+
+			// per ZK docs, this watcher should only trigger once
+			client.checkExists().usingWatcher(actualWatcher).forPath("/test");
+			client.getData().usingWatcher(actualWatcher).forPath("/test");
+
+			client.setData().forPath("/test", "foo".getBytes());
+			client.delete().forPath("/test");
+			timing.sleepABit();
+			assertEquals(actualWatcherCount.getAndSet(0), 1);
+
+			client.create().forPath("/test");
+			client.checkExists().usingWatcher(actualWatcher).forPath("/test");
+			client.delete().forPath("/test");
+			timing.sleepABit();
+			assertEquals(actualWatcherCount.get(), 1);
+		} finally {
+			CloseableUtils.closeQuietly(client);
+		}
+	}
+
+	@Test
+	public void testSetAddition() {
+		Watcher watcher = new Watcher() {
+			@Override
+			public void process(WatchedEvent event) {
+
+			}
+		};
+		NamespaceWatcher namespaceWatcher1 = new NamespaceWatcher(null, watcher, "/foo");
+		NamespaceWatcher namespaceWatcher2 = new NamespaceWatcher(null, watcher, "/foo");
+		assertEquals(namespaceWatcher1, namespaceWatcher2);
+		assertFalse(namespaceWatcher1.equals(watcher));
+		assertFalse(watcher.equals(namespaceWatcher1));
+		Set<Watcher> set = Sets.newHashSet();
+		set.add(namespaceWatcher1);
+		set.add(namespaceWatcher2);
+		assertEquals(set.size(), 1);
+	}
+
+	@Test
+	public void testCuratorWatcher() throws Exception {
+		Timing timing = new Timing();
+		// Create variables for tracking behaviors of mock object
+		AtomicInteger watcherCount = new AtomicInteger(0);
+		// Construct mock object
+		CuratorWatcher watcher = mock(CuratorWatcher.class);
+		// Method Stubs
+		doAnswer((stubInvo) -> {
+			watcherCount.incrementAndGet();
+			return null;
+		}).when(watcher).process(any(WatchedEvent.class));
+		CuratorFramework client = CuratorFrameworkFactory.newClient(server.getConnectString(), timing.session(),
+				timing.connection(), new RetryOneTime(1));
+		try {
+			client.start();
+			client.create().forPath(PATH);
+			// Add twice the same watcher on the same path
+			client.getData().usingWatcher(watcher).forPath(PATH);
+			client.getData().usingWatcher(watcher).forPath(PATH);
+			// Ok, let's test it
+			client.setData().forPath(PATH, new byte[] {});
+			timing.sleepABit();
+			assertEquals(1, watcherCount.get());
+		} finally {
+			CloseableUtils.closeQuietly(client);
+		}
+	}
+
+	@Test
+	public void testZKWatcher() throws Exception {
+		Timing timing = new Timing();
+		CountZKWatcher watcher = new CountZKWatcher();
+		CuratorFramework client = CuratorFrameworkFactory.newClient(server.getConnectString(), timing.session(),
+				timing.connection(), new RetryOneTime(1));
+		try {
+			client.start();
+			client.create().forPath(PATH);
+			// Add twice the same watcher on the same path
+			client.getData().usingWatcher(watcher).forPath(PATH);
+			client.getData().usingWatcher(watcher).forPath(PATH);
+			// Ok, let's test it
+			client.setData().forPath(PATH, new byte[] {});
+			timing.sleepABit();
```

Comment:
What about using Awaiatility and wait for the condition to become 'true'

Reply:
Make sense to me, thanks for the suggestion, I'll use Awaiatility to replace timeing.